src/compiler/emitter.ts

```diff
@@ -144,81 +178,197 @@ module ts {
+                    return !!(predefinedTempsInUse & TempVariableKind._i);
+                }
+
+                if (name === "_n") {
+                    return !!(predefinedTempsInUse & TempVariableKind._n);
+                }
+
+                if (name.length === 2 && name.charCodeAt(1) >= CharacterCodes.a && name.charCodeAt(1) <= CharacterCodes.z) {
+                    // handles _a .. _z
+                    let n = name.charCodeAt(1) - CharacterCodes.a;
+                    return n < tempCount;
                 }
                 else {
-                    currentScopeNames = undefined;
+                    // handles _1, _2...
+                    let n = +name.substring(1);
+                    return !isNaN(n) && n >= 0 && n < (tempCount - 26);
                 }
             }
 
-            function generateUniqueNameForLocation(location: Node, baseName: string): string {
-                let name: string
-                // first try to check if base name can be used as is
-                if (!isExistingName(location, baseName)) {
-                    name = baseName;
+            // This function generates a name using the following pattern:
+            // _a .. _h, _j ... _z, _0, _1, ...
+            // It is guaranteed that generated name will not shadow any existing user-defined names,
+            // however it can hide another name generated by this function higher in the scope.
+            // NOTE: names generated by 'makeTempVariableName' and 'makeUniqueName' will never conflict.
+            // see comment for 'makeTempVariableName' for more information.
+            function makeTempVariableName(location: Node, tempVariableKind: TempVariableKind): string {
+                let tempName: string;
+                if (tempVariableKind !== TempVariableKind.auto && !(predefinedTempsInUse & tempVariableKind)) {
+                    tempName = tempVariableKind === TempVariableKind._i ? "_i" : "_n";
+                    if (resolver.isUnknownIdentifier(location, tempName)) {
+                        predefinedTempsInUse |= tempVariableKind;
+                        return tempName;
+                    }
                 }
-                else {
-                    name = generateUniqueName(baseName, n => isExistingName(location, n));
+
+                do {
+                    // Note: we avoid generating _i and _n as those are common names we want in other places.
+                    var char = CharacterCodes.a + tempCount;
+                    if (char !== CharacterCodes.i && char !== CharacterCodes.n) {
+                        if (tempCount < 26) {
+                            tempName = "_" + String.fromCharCode(char);
+                        }
+                        else {
+                            tempName = "_" + (tempCount - 26);
+                        }
+                    }
+
+                    tempCount++;
+
+                } while (!resolver.isUnknownIdentifier(location, tempName));

```

Comment:
Yes, but I don't really like the terms known/unknown. I know what the identifier is, because I see the identifier. Maybe something like resolvesToSomeSymbol.